rfcs/1111-first-class-virtual-device.md

```diff
@@ -0,0 +1,146 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The AST change will occur in include/tvm/ir/expr.h:
+
+```
+class RelayExprNode : public BaseExprNode {
+ public:
+  /*!
+   * \brief Stores the result of type inference(type checking).
+   *
+   * \note This can be undefined before type inference.
+   *       This value is discarded during serialization.
+   */
+  mutable Type checked_type_ = Type(nullptr);
+  /*!
+   * \return The checked_type
+   */
+  inline const Type& checked_type() const;
+  /*!
+   * \brief Check if the inferred(checked) type of the Expr
+   *  is backed by a TTypeNode and return it.
+   *
+   * \note This function will thrown an error if the node type
+   *       of this Expr is not TTypeNode.
+   *
+   * \return The corresponding TTypeNode pointer.
+   * \tparam The specific TypeNode we look for.
+   */
+  template <typename TTypeNode>
+  inline const TTypeNode* type_as() const;
+
+  /*!
+   * \brief The virtual device (SEScope) for this node (the result of device planning).
+   *
+   * \note Unfortunately, the type of virtual_device_ needs to be ObjectRef to avoid a circular import.
+   *       We can forward-declare the SEScope type for the getter function, but not for the field
+   *       itself.
+   */
+  mutable ObjectRef virtual_device_;
+
+  /*!
+   * \return The virtual device (currently called SEScope, this will be changing soon.)
+   */
+  SEScope virtual_device() const;
+
+  static constexpr const char* _type_key = "RelayExpr";
+  static constexpr const uint32_t _type_child_slots = 22;
+  TVM_DECLARE_BASE_OBJECT_INFO(RelayExprNode, BaseExprNode);
+};
+```
+
+Additionally, I will add virtual_device_ to the WithFields methods.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One challenge with making virtual devices first class is that passes in TVM do not propagate all fields when they visit expressions. You can see this today with spans-- most visitors do not preserve spans. When we introduce the virtual device field, we will need to ensure that it is propagated correctly throughout the Relay program. To do this, I introduced WithFields (code: https://github.com/apache/tvm/blob/main/src/relay/ir/expr.cc#L79-L99), a COW constructor that copies extra fields, including spans.  I will extend WithFields to also copy virtual devices. Then, we can use WithFields to ensure that the virtual device field is correctly propagated. 
+
+Additionally, passes that move or introduce expressions will need to cooperate with device annotation (i.e., figure out the correct device themselves and insert it in any expressions they create), or device planning will need to be run again after the pass (similar to the type inference pass). In the future, we'd like to introduce a 'lite' version of the device planning pass that "fills in the gaps" left by these passes. It will propagate virtual device information strictly upwards and will have much less overhead than the full device planning pass.
```

Comment:
IIUC, after this RFC `virtual_device_` behaves almost the same as `checked_type_`, so this requirement makes sense to me. Meanwhile, I'm not sure how frequent we should run this new pass to preserve/propagate `virtual_device_`. If we have to run it as frequent as `InferType`, one thought I'm having is whether it is reasonable to treat first class `virtual_device_` as a part of type, and directly implement the virtual device propagation logic in `InferType`; otherwise we may need to run much more passes which may further slow down the compilation.

Reply:
Currently the domains of types and virtual devices are orthogonal-- for example, tuples can only have one virtual device but each field of a tuple has a type. There's more context on the connection between types and virtual devices in device_domains.h and in device_planner.cc if you are curious. 

@mbs-octoml could you expand on this point a bit?

Reply:
Yeah, that's an excellent question Cody. Currently the 'device domain' used during PlanDevices is
  D ::=  <device> | fn(D, ..., D) -> D
So it's *slightly* type-dependent but otherwise very 'forgetful'. I think that's enough to justify keeping devices out of types, and their inference separate.

HOWEVER there are plenty of attributes which are refinements on types, and I suspect we'll want to put them in the type nodes directly. But even then I'd still not lump their analysis with type checking. That is, I'd still write the analysis as a stand alone pass, but the pass may make liberal use of induction on the already known types.

Would very  much like to keep an open mind on this one and use a concrete example to test the water. Maybe side effects. Maybe uniqueness (for in-place updates).

Reply:
Thanks you both for the clarification. Lily pointed out a good point that I didn't notice and Mark made it clearer. I agree that we could make it a standalone pass first and do more optimizations to shorten the compile time later if necessary. And definitely, side effects and in-place updates might bring many new issues in the future, but I believe this design should cover them in general :D